src/Iterators.jl

```diff
@@ -894,4 +895,94 @@ macro chain(ex)
     Expr(:let, Expr(:block, cycleex...), states...)
 end
 
+
+type MergedIter
+    it1   # iterator 1
+    it2   # iterator 2
+    lt    # custom `<` operator to compare elements of 2 iterators    
+end
+
+MergeIter(it1, it2) = MergeIter(it1, it2, isless)
+
+type MergedState
+    hd1   # current head of 1st iterator
+    hd2   # current head of 2nd iterator
+    s1    # current state of 1st iterator
+    s2    # current state of 2nd iterator
+end
+
+
+function start(it::MergedIter)
+    s1 = start(it.it1)
+    s2 = start(it.it2)
+    # doing 1 step ahead and caching next element of each iter
+    hd1, s1 = safenext(it.it1, s1)
+    hd2, s2 = safenext(it.it2, s2)
+    return MergedState(hd1, hd2, s1, s2)
+end
+
+"""
+Like `next`, but may be called on iterator that is already done. In this case
+new item is `nothing` and new state is the same as previous. 
+"""
+function safenext(it, s)
+    if done(it, s)
+        return nothing, s
+    else
+        return next(it, s)
+    end
+end
+
+function next(it::MergedIter, s::MergedState)
+    lt = it.lt
+    if done(it, s) throw(ArgumentError("Iterator is done")) end
```

Comment:
I don't really have a strong opinion here, but I'm curious what's the conventional approach to handle `next()` after `done(...) == true`? 

I know that for objects like ranges we can just continue producing new elements. However, for IO-based iterators (e.g. iterator over a file) we can't really produce anything else after the end of EOF. So what is supposed to be an appropriate result for calls to `next()` for such iterators? 

(In fact, I'm working on one such iterator right now, so would be glad to make it using best practices)